src/main/java/picard/illumina/parser/readers/CbclReader.java

```diff
@@ -57,117 +63,189 @@
  * higher 4 bits are the second cluster.
  **/
 
-public class CbclReader extends BaseBclReader implements CloseableIterator<BclData> {
+public class CbclReader extends BaseBclReader implements CloseableIterator<CbclData> {
 
-    private static Log log = Log.getInstance(CbclReader.class);
-
-    private ByteBuffer[] cachedTile;
-
-    private int[] currentTile;
-
-    private List<BclData> queue = new ArrayList<>();
+    private final byte[][] cachedTile;
+    private final int[] cachedTilePosition;
 
+    private CbclData queue = null;
+    private Iterator<AbstractIlluminaPositionFileReader.PositionInfo> positionInfoIterator;
     private final CycleData[] cycleData;
-    private Map<Integer, File> filters;
-    private Map<Integer, boolean[]> cachedFilter = new HashMap<>();
-
-    void addFilters(Map<Integer, File> filters) {
-        this.filters = filters;
-    }
+    private final Map<Integer, File> filterFileMap;
+    private final Map<Integer, boolean[]> cachedFilter = new HashMap<>();
+    private final Map<Integer, Map<Integer, File>> surfaceToTileToCbclMap;
+    private int headerSize;
+    private Map<Integer, List<TileData>> allTiles = new HashMap<>();
 
     private static final int INITIAL_HEADER_SIZE = 6;
+    private static final Log log = Log.getInstance(CbclReader.class);
 
-    public CbclReader(final List<File> cbcls, final int[] outputLengths) {
+    public CbclReader(final List<File> cbcls, final Map<Integer, File> filterFileMap, final int[] outputLengths,
+                      int tileNum, List<AbstractIlluminaPositionFileReader.PositionInfo> locs, boolean headerOnly) {
         super(outputLengths);
+        surfaceToTileToCbclMap = sortCbcls(cbcls);
+        this.filterFileMap = filterFileMap;
         cycleData = new CycleData[cycles];
-        cachedTile = new ByteBuffer[cycles];
-        currentTile = new int[cycles];
+        cachedTile = new byte[cycles][];
+        cachedTilePosition = new int[cycles];
+        for (int i = 1; i <= cycles; i++) {
+            allTiles.put(i, new ArrayList<>());
+        }
+        readSurfaceTile(tileNum, locs, headerOnly);
+        close();
+    }
 
+    private void readSurfaceTile(int tileNum, List<AbstractIlluminaPositionFileReader.PositionInfo> locs,
```

Comment:
One thing I noticed during yesterday's talk was that this is a very long method, which makes it hard to see what's going on here. It would be easier to understand if it could be split into three or so methods.